### ISSUE TITLE
Cannot use custom protocol with payload

### ISSUE THREAD

**Issue by maruu** [seladb/PcapPlusPlus#496]
Hi,

I have a custom protocol that has a header and a payload. Until now I have implemented the header as its own layer and then added a `PayloadLayer` with the payload. The custom protocol allows chaining of multiple header/payload pairs in one Ethernet frame. Implementing this functionality raised a problem with the current implementation of `insertLayer`.
If I add a second header layer after the `PayloadLayer` of the first header/payload pair, the steps regarding setting the `m_Data` ptr and `m_DataLen` won't work, since the `PayloadLayer` returns `m_DataLen` on `getHeaderLen()` - this effectively consumes all data that it is in the Packet buffer, although it may not be part of the `PayloadLayer` anymore.

After taking a look at the public interface of the Layer baseclass, I noticed functions for getting a LayerPayload (`getLayerPayload` and `getLayerPayloadSize`) - although, with the current implementation of the `insertLayer` method, I cannot think of a way how adding the payload to my layer directly instead of adding a `PayloadLayer` would solve this problem. This is because `insertLayer` only uses `getHeaderLen()`.

We already talked about this - sort of - inconsistency when I raised a bug report showing that `getLayerPayloadSize()` called on `PayloadLayer` returns 0. You told me back then that this was exactly because of this `insertLayer` dilemma.

One option would be to implement a Layer class that incorporates both header and payload data and that would return the complete length on `getHeaderLen()` call, but I feel that this is just a way to avoid a problem, and not a real solution.

My question is now: How would you go about implementing this protocol with header/payload pair as well as the chaining of the protocol in one Ethernet frame. Is this even possible with the current state of PcapPlusPlus?

I am looking forward to hearing from you.

Thanks,
maruu

**Comment by seladb** [comment]
hi @maruu , thanks for raising this issue. I don't have the details of your custom protocol, but if your protocol chains multiple messages on the same Ethernet frame I'd actually recommend having both the header and the payload part of the Layer object. That would help distinguishing between the different messages.

A somewhat similar example is the SSL/TLS protocol which also allows chaining multiple messages on a single Ethernet frame. The way I implemented this protocol is by encapsulating the payload inside the message, as you can see in [`SSLApplicationDataLayer`](https://github.com/seladb/PcapPlusPlus/blob/a3504afc29c470abe5b19989e405941a6e79e3ce/Packet%2B%2B/header/SSLLayer.h#L447) for example

**Comment by maruu** [comment]
Thank you for your answer.
What advantages do you see to implement the protocol as you described, creating one layer for the header and one for the payload? With this design a user would be able to create a payload layer without a header. From my perspective this design choice would allow wrong use of the layers? 
If one would implement the header and payload pair in one layer and - I am just thinking out loud now - would extend the insertLayer (and others) ptr and length resetting parts (as can be seen in the following: )

https://github.com/seladb/PcapPlusPlus/blob/a3504afc29c470abe5b19989e405941a6e79e3ce/Packet%2B%2B/src/Packet.cpp#L279-L298

by just adding the LayerPayloadSize to the calculation of the next start of the layer, which is currently calculated here:

https://github.com/seladb/PcapPlusPlus/blob/a3504afc29c470abe5b19989e405941a6e79e3ce/Packet%2B%2B/src/Packet.cpp#L292-L294

and extending it to look like this:
```
		// advance data ptr and data length
		dataPtr += curLayer->getHeaderLen() + curLayer->getLayerPayloadSize();
		dataLen -= curLayer->getHeaderLen() + curLayer->getLayerPayloadSize();
```

If the layer has got no payload, the behaviour is as it has been until now. Even with the `PayloadLayer`, which returns 0 on `getLayerPayloadSize()`.

So, to summarize, what I would propose for such protocols would be to use the already existing `getLayerPayload()` interface of the `Layer` base class and extend the dataPtr reset algorithms used in the `Packet` class to take this into account as described above. The header and payload logic would then be coupled in one Layer class, which would prevent users from creating payload layers without preceding header layers.

Please let me know what you think.

Thanks,
maruu

**Comment by maruu** [comment]
Thank you for your answer.
What advantages do you see to implement the protocol as you described, creating one layer for the header and one for the payload? With this design a user would be able to create a payload layer without a header. From my perspective this design choice would allow wrong use of the layers?

Yesterday I have tried my luck at implementing my needed functionality and I came up with the following:

I have united both header and payload in one layer. This layer returns its complete length (inc. payload length) on a call to `getHeaderLen()`. I then proceeded to declare the base Layer class' `getLayerPayload()` and `getLayerPayloadSize()` methods as virtual and overrid them in my layer to return only the payload, and not a calculation of `m_Data + getHeaderLen()` as does the current base Layer class implementation.

I am now able to chain multiple header/payload pairs in one Ethernet frame with this approach. What I dislike about this approach is the "abuse" of the `getHeaderLen()`. But I have also already raised this with the issue that I have created regarding the unexpected return value of `PayloadLayer` for a call to `getLayerPayloadSize()`. I am not sure how it came to this, but maybe we can streamline the internals a bit to avoid wrong use of Layers, as was the case on my side when I tried to use the PayloadLayer :-)

I have also taken a look at your SSLLayer implementation and I now understand what you meant by

> That would help distinguishing between the different messages.

although this is not a use-case that I have, since I treat the payload as payload - meaning that I don't interpret it. The payload usually consists of dynamic arrays and other complex types, whose structure is usually only known to the application that receives the data. So I don't think that parsing of the payload - in my case - might be a use-case for PcapPlusPlus. In terms of the SSLLayer, this of course makes sense.

I am curious about your thoughts regarding my implementation and maybe you have an idea what we can do about the "abuse" of the `getHeaderLen()` function.

**Comment by seladb** [comment]
> What advantages do you see to implement the protocol as you described, creating one layer for the header and one for the payload? With this design a user would be able to create a payload layer without a header. From my perspective this design choice would allow wrong use of the layers?

I actually suggest the opposite - to create one layer that contains both the header and the payload.

> I then proceeded to declare the base Layer class' `getLayerPayload()` and `getLayerPayloadSize()` methods as virtual and overrid them in my layer to return only the payload, and not a calculation of `m_Data + getHeaderLen()` as does the current base Layer class implementation.

I don't think you should override these methods. I do agree that it's seems a little confusing, but the purpose of `getHeaderLen()` is to get the layer data length, and the purpose of `getLayerPayloadSize()` is to get its payload. If we take IPv4 as an example then the header is the IPv4 header (`ip_hdr`) and the payload is everything that comes after it, including other layers like TCP, HTTP, etc. Same goes for other protocols like IPv6, TCP, UDP, etc.

Your use case (which is somewhat similar to SSL/TLS) is a little different because a single Ethernet frame can contain multiple messages. So in these cases "header" and "payload" get a different meaning: "header" is actually a complete message (which in your case includes both a header and a payload) and "payload" includes all the messages the comes after. So `getHeaderLen()` should return the message length, and `getLayerPayloadSize()` should return the length of data that comes after the message (which is exactly what the current implementation does). Inside of your layer you should implement methods that retrieve the header and payload of your message. 

I do understand it's confusing but I don't see any other design that would better support these 2 use-cases. If you have any suggestion please don't hesitate to propose.

**Comment by maruu** [comment]
Hey @seladb,

sorry for taking so long to get back to you. I understand the correct usage of the ``getHeaderLen()`` and ``getLayerPayloadSize()`` better now. I put it on my to-do list to get back to my implementation and change it to the way you have just clarified. Of course, as this library is already heavily used in several projects around the Internet, I guess it is too late to change the ``getHeaderLen()`` function name to some name that expresses the actual intent more clearly, e.g. ``getLayerSize()``? :-)

Anyway, thanks for your time and your thoughtful comments, as always. :-)

If you want, we can close this issue.

**Comment by seladb** [comment]
Sure, thanks!

I'm closing the issue now
